### ISSUE TITLE
Different Results in Subsequent Runs of Same Pipeline

### ISSUE THREAD

**Issue by mchildsCO76** [openMVG/openMVG#1128]
We have noticed getting different results from an OpenMVG pipeline running the exact same data through on the same machine, with placement differences of a meter or more in the final point cloud. 

In comparing the detailed logs everything looks identical until IncrementalSfM is running. A comparison of the logs shows first some very small differences in reported precisions during Robust Resection, then more and more differences.

What could cause differences when running the exact same data pipeline on the same machine? I would expect the same inputs to always produce the same outputs, or at least something so close that the difference is academic due to something like thread timing processing one data set before another.

I have attached log files for 3 subsequent runs and a screenshot demonstrating the differences in the point locations along a small section near the center of the drone image area. The logs list all of the input images, sizes, and locations, as well as the complete list of run commands.

[run2_log.txt](https://github.com/openMVG/openMVG/files/1420121/run2_log.txt)
[Run3_log.txt](https://github.com/openMVG/openMVG/files/1420122/Run3_log.txt)
[run1_log.txt](https://github.com/openMVG/openMVG/files/1420123/run1_log.txt)
![profile_difference](https://user-images.githubusercontent.com/29988262/32077855-0ecf083c-ba62-11e7-8688-7eaf4ac71638.png)

**Comment by pmoulon** [comment]
Using multithreading makes some operation performed in a different order and so the result can be a bit different.

If you want to obtain the same result every time, you have to set one thread with the OMP_NUM_THREAD variable and you will obtain a reproducible result.

**Comment by mchildsCO76** [comment]
Is there any way to maintain use of multiple threads while still converging on the best solution? Ideally the result would be the optimal solution regardless of how the processing was divided up.

**Comment by pmoulon** [comment]
>  Ideally the result would be the optimal solution regardless of how the processing was divided up.

Sure, I would like to make this true too.
Feel free to help on this (I mean track where the difference appears in order to help us improve the repeatability in a multithread setting)

**Comment by mchildsCO76** [comment]
What can I add to help track where it diverges?

**Comment by mchildsCO76** [comment]
Sorry thought I posted that a week ago but it was sitting there ready for me to press 'Comment' :)

**Comment by pmoulon** [comment]
Sorry for the late answer.

You can compare two logs and see where it starts to be different.

**Comment by mchildsCO76** [comment]
That's what the first comment is. I attached logs from two separate runs on the same data. The logs are the same (other than timestamps and such) until the IncrementalSfM step, then they start to differ. I tried changing a few threading model things on a wild hunch, but they didn't help. I don't know enough about how the threading in IncrementalSfM works to know where to start looking for the differences between runs making the results non-deterministic.
